Common/src/linear_solvers_structure.cpp

```diff
@@ -657,52 +728,54 @@ unsigned long CSysSolve::Solve(CSysMatrix & Jacobian, CSysVector & LinSysRes, CS
   }
 
   /*--- Solve the linear system using a Krylov subspace method ---*/
-  
+
+  HandleTemporariesIn(LinSysRes, LinSysSol);
+
   if (KindSolver == BCGSTAB || KindSolver == CONJUGATE_GRADIENT ||
       KindSolver == FGMRES  || KindSolver == RESTARTED_FGMRES ) {
     
-    mat_vec   = new CSysMatrixVectorProduct(Jacobian, geometry, config);    
-    CPreconditioner* precond = NULL;
+    mat_vec = new CSysMatrixVectorProduct<ScalarType>(Jacobian, geometry, config);
+    CPreconditioner<ScalarType>* precond = NULL;
     
     switch (KindPrecond) {
       case JACOBI:
         Jacobian.BuildJacobiPreconditioner();
-        precond   = new CJacobiPreconditioner(Jacobian, geometry, config);
+        precond = new CJacobiPreconditioner<ScalarType>(Jacobian, geometry, config);
         break;
       case ILU:
         Jacobian.BuildILUPreconditioner();
-        precond   = new CILUPreconditioner(Jacobian, geometry, config);
+        precond = new CILUPreconditioner<ScalarType>(Jacobian, geometry, config);
         break;
       case LU_SGS:
-        precond = new CLU_SGSPreconditioner(Jacobian, geometry, config);
+        precond = new CLU_SGSPreconditioner<ScalarType>(Jacobian, geometry, config);
         break;
       case LINELET:
         Jacobian.BuildJacobiPreconditioner();
-        precond = new CLineletPreconditioner(Jacobian, geometry, config);
+        precond = new CLineletPreconditioner<ScalarType>(Jacobian, geometry, config);
         break;
       default:
         Jacobian.BuildJacobiPreconditioner();
-        precond = new CJacobiPreconditioner(Jacobian, geometry, config);
+        precond = new CJacobiPreconditioner<ScalarType>(Jacobian, geometry, config);
         break;
     }
     
     switch (KindSolver) {
       case BCGSTAB:
-        IterLinSol = BCGSTAB_LinSolver(LinSysRes, LinSysSol, *mat_vec, *precond, SolverTol, MaxIter, &Residual, ScreenOutput);
+        IterLinSol = BCGSTAB_LinSolver(*LinSysRes_ptr, *LinSysSol_ptr, *mat_vec, *precond, SolverTol, MaxIter, &Residual, ScreenOutput);
```

Comment:
Same here.. might be a conflict w #652 since I now send in the config container at the end of the argument list. Should be minor.